aws-rds-dbclusterparametergroup/src/main/java/software/amazon/rds/dbclusterparametergroup/BaseHandlerStd.java

```diff
@@ -238,57 +254,159 @@ private ProgressEvent<ResourceModel, CallbackContext> validateModelParameters(
         return progress;
     }
 
-    private ProgressEvent<ResourceModel, CallbackContext> describeCurrentDBClusterParameters(
+    private Iterable<Parameter> fetchDBClusterParametersIterable(
+            final ProxyClient<RdsClient> proxyClient,
+            final DescribeDbClusterParametersRequest request
+    ) {
+        Iterable<Parameter> result = Collections.emptyList();
+
+        String marker = null;
+        int page = 0;
+        do {
+            if (page >= MAX_DESCRIBE_PAGE_DEPTH) {
+                throw new CfnInvalidRequestException("Max DescribeDBParameters page reached.");
```

Comment:
Not sure why this `CfnInvalidRequestException` and not `RuntimeException` like fetch `DefaultEngineVersion`. This error not dependent on customer input. Customer can not trigger this error via his stack.

Reply:
@moataz-mhmd regarding the iterables: well, `Iterable` is just an interface, not an implementation. The reason is is kept in is because it allows me to concatenate iterable objects together and hide the details of the pagination. Paginator operations are actually a good help when dealing with large amount of paginated requests, but there is a significant flaw that forced me to avoid using them here in DBClusterParameterGroup and in DBParameterGroup. See https://github.com/aws-cloudformation/cloudformation-cli-java-plugin/issues/414 for more details. Once this ticket is fixed, I would prefer to remove these custom `do while` loops and return back to the paginated requests. Keeping `Iterable` in allows us to replace this code with minimal efforts.
I'm not sure I understand the memory consumption concerns you shared. As I've mentioned, `Iterable` does not dictate the use of a specific of the underlying data structure in the container implementing this interface. In our case the underlying container is `List<Parameter>`. I have serious doubts that a `HashMap` would be a more efficient storage memory-wise and GC-wise. Could you please provide a more specific data to back this concern?

Reply:
I am not sure how a bad request from RDS API's perspective relates to `CfnInvalidRequestException`. Customer will get `CfnInvalidRequestException` not `RdsInvalidRequest`

Reply:
Got it. You want to keep it refactored in way once the bug resolved we rewrite it again with paginators.

Reply:
One more note: Runtime exception needs to be mapped to CFN error code. I think throwing very generic exception can prevent us from doing so. `CfnInvalidRequestException` is by default mapped to `InvalidRequest` and terminate, meanwhile` RuntimeException` will be converted into `InternalError` and retried.

Therefore I think `CfnInvalidRequestException` serves our purposed better at this point in time.